### ISSUE TITLE
ControlGUI

### ISSUE THREAD

**Issue by JE521** [SammyIAm/Moppy2#76]
Hi, I know moppy is really old, but i was interested. I am on pc. I followed the directions on the wiki, but when I am supposed to open the "MoppyControlGUI.bat", (I was given the prompt to agree to run it) , it opens for a millisecond and then closes. The Hardware and arduino set up worked fine, and i am using moppy 2.0.1. If you could help that would be much appreciated. Thank you

**Comment by SammyIAm** [comment]
Replying by email so excuse the brevity (I can try more later if this
doesn't help):

Instead of double clicking the .bat, try opening the command line
(Start>run>cmd) and running the bat from there.  It'll show any errors that
occur hopefully.

Also look around in the Moppy folders for log files that might have errors.

Maybe try before unzipping, right clicking, properties, and check
"unblock", then unzip.

If it's still not working, or there's an error, let me know and I'll
respond at length later.

Good luck!

**Comment by JE521** [comment]
Thank you for responding. I copied the text contents of the .bat file and pasted it in command prompt. I hop thats what you meant. but it did the same thing. I also unblocked it. when i clicked on it, it would pull up a window and had the options "extract all", "Run", and "Cancel". I clicked run, did the same thing. I extracted all and it just made another duplicate moppycontrolGUI folder. 

I also noticed in the wiki, it has no mention of netbeans, is netbeans not required anymore?

**Comment by SammyIAm** [comment]
Yup, 5.25" drives work just fine (though some of them may have less tracks, so you'll need to adjust the maximum position for those drives in the Arduino code (to 79 instead of 158)).

You can add up to 9 floppies by just connecting them to the next available pair of pins (e.g. 2&3, 4&5, &c.).

**Comment by JE521** [comment]
Do you have a pinout diagram by chance? I can't seem to find it in its manual.

**Comment by SammyIAm** [comment]
5.25" drives, I believe, have the same pin ordering as the 3.5" drives, just a different connector type.

**Comment by JE521** [comment]
I am working on the connections, but how do I change the max position for any of the drives. I checked the code and couldn't find any variable for the max position. Also is it ok to connect all of the drive's ground pins to the same ground pin on the arduino. I doubt it but I don't see much different ground pins on the arduino.

**Comment by JE521** [comment]
the drive does not work, is it possible to control the stepper by directly connecting it to the arduino. I looked online and in your previous issues but can not find anything. Thanks again so much for your guidance.

**Comment by JE521** [comment]
I also can not seem to run the working floppy on any of the other pins for additional drives, i was checking to see if it worked and it didn't.

**Comment by SammyIAm** [comment]
GitHub has a [useful search function](https://github.com/SammyIAm/Moppy2/search?q=max+position&unscoped_q=max+position) to find things in a repository.

It's not possible to control the stepper motor directly from the Arduino (there is some circuitry required to drive stepper motors, but it's conveniently included in the floppy drive hardware which is why floppy drives work so well).  It *is* possible to drive stepper motors directly using Moppy, but you'd need a driver board connected to the Arduino to do so.

Make sure the MIDI file you're using has enough tracks for the other pins to have things happening on them, and make sure your mapping settings in the GUI are sending events for all the MIDI channels.

**Comment by DarianKimberly** [comment]
Is there a linux version or will I need to completely rewrite the code in a compatible format?

**Comment by SammyIAm** [comment]
@DarianKimberly , Moppy's written in Java so you're good to go pretty much anywhere.  In the `bin` directory there's a file named `MoppyControlGUI` that *should* (I haven't actually tested) allow you to easily run the ControlGUI.  If that script doesn't work out of the box, it should only require minor adjustments, or you should be able to run `MoppyControlGUI-2.0.1.jar` directly as long as you set the classpath correctly.

**Comment by DarianKimberly** [comment]
Not sure how to set the classpath. New to this.

**Comment by SammyIAm** [comment]
Did the script not work?  What error are you seeing when you run it?

I'm happy to try to help answer specific questions; but because there are a ton of different variables to a given setup and Moppy generally just uses pretty standard libraries, you'll probably have just as much (or better) luck on Google for a lot of the generic getting-Java-working-on-my-system questions.

**Comment by DarianKimberly** [comment]
Just trying to figure out if I can control my arduino from my Raspbian
setup as it is linux based. Not sure if it is compatible. the controller
program is a windows batch file, and im not sure how to convert it into
something linux friendly.

**Comment by DarianKimberly** [comment]
The reason behind this is that I want to make a portable version of the
moppy built that i can control wirelessly through a raspberry pi zero w
connected to my uno and VNC viewer.

**Comment by SammyIAm** [comment]
So, just to clarify again, the controller program is a `.jar` file.  There is a Windows .bat file included for running on Windows machines, **and a shell script** (the file I mentioned before in the `bin` directory) for running on Linux/Unix.

Try the shell script.

**Comment by DarianKimberly** [comment]
I tried and it just opens as a text file

**Comment by DarianKimberly** [comment]
Update. I manually set it to executable (for some reason this is not a
default), and it opens for a brief moment, then just closes.

**Comment by SammyIAm** [comment]
Try executing it from a terminal.  It should output any errors in encounters (you can also look for `*.log` files in the Moppy directories).

If I had to guess though, it's probably a `JAVA_HOME` environment variable issue.  Make sure that's set up correctly.  (But try running from a terminal to look for the error first, it might be something else)

**Comment by DarianKimberly** [comment]
Again, no dice. I already tried executing it from the terminal.

**Comment by SammyIAm** [comment]
What was the output when you tried to run it?  (I can't help if I don't know what issues you're running into)

**Comment by DarianKimberly** [comment]
Executing in terminal results in a terminal that opens for a split second
and immediately closes. Nothing else happens.

**Comment by JE521** [comment]
you could try the JAVA_HOME thing, just scroll up to beginning.

**Comment by JE521** [comment]
I think that the pin problem might be due to my use of an arduino mega 2560 instead of an uno. I will try to purchase an uno if there is no way to do this with my board. Also I will be less active on github because of school.

**Comment by DarianKimberly** [comment]
What do you mean by that?

**Comment by SammyIAm** [comment]
@JE521 , The Arduino Mega should work fine for the first set of pins (roughly 2-19 or so).  Can you maybe send a screenshot of your ControlGUI, just in case something with the mappers is preventing channels higher than 1 from being sent to the drives?

No rush of course, enjoy (?) school.  ;P

**Comment by SammyIAm** [comment]
@DarianKimberly .  JE521 is referring to the original topic of this issue which he created.

Generally it's best to create a new issue if you're asking for help with an unrelated problem.

**Comment by DarianKimberly** [comment]
The problem is getting the GUI program to open. It just opens a command
line for an instant, then immediately closes.

**Comment by SammyIAm** [comment]
@DarianKimberly : I was talking to JE521, not you.  Why don't you open up a new issue and we'll see if we can't get you sorted out over there.

**Comment by DarianKimberly** [comment]
Sorry. My bad. It emails me with every reply.

**Comment by JE521** [comment]
![Moppy Control 9_13_2019 5_19_27 PM](https://user-images.githubusercontent.com/53789946/64895496-c4408f00-d64a-11e9-8a07-57a7de79417d.png)

**Comment by JE521** [comment]
Also I have a driver board for my stepper.

**Comment by SammyIAm** [comment]
Alright, so... this might seem like a silly suggestion, but I figure we should just make sure...  The MIDI file you have open is called `KirbysTheme (1 floppy).mid` which I assume has been written for just one floppy drive (and therefore probably doesn't contain any notes for any of the other drives.

As something to try though, you can click "Add Mapper" and set up the mappers like this:

| | | | |
|---|---|---|---|
|`n>22 && n<67`|`1`|`1`|`n`|
|`n>22 && n<67`|`1`|`2`|`n`|

That will basically take any note that comes in and assign it to SubAddress 1 (first row) and SubAddress 2 (second row).  That'll make the first two floppy drives play the same notes, but could help testing your second drive.

Regarding the stepper motor, as long as the board is sufficiently similar to the L298N board, you should be able to comment out the two FloppyDrives lines in `MoppyCore.cpp` and uncomment the L298N lines.  The L298N code was contributed by someone else so I don't know the nitty gritty, but if it doesn't work out of the box it should work with some small modifications.

**Comment by JE521** [comment]
That sounds like a good idea, but the 4,5 pins and so on don't work, I try restarting arduino on those pins and floppy wont play startup sound. Also the subaddress says play for 1 floppy because I set that when I saw the issue you forwarded me.

**Comment by JE521** [comment]
Also the 5.5" floppy doesn't work, so I am trying to tap into its stepper motor. And I am getting more 3.25" floppies soon.

**Comment by SammyIAm** [comment]
- The startup sound *only* plays on the first drive.  (You could change it in the [Arduino code here](https://github.com/SammyIAm/Moppy2/blob/master/Arduino/Moppy/src/MoppyInstruments/FloppyDrives.cpp#L85) if you wanted to)
- The **MIDI file** says "1 floppy" (try a different MIDI file!)
- You should be able to figure out how to get the 5.25" floppy working by manually grounding pins to see which ones control the stepping, &c.  I guess it's possible the drive itself is dead and then you could just bypass it and control the motor directly.

**Comment by JE521** [comment]
I hooked up the right pins to the 2,3 pins on arduino. it wont move.

**Comment by JE521** [comment]
I am sorry i cut off this issue without notice, my computer broke and I was busy with school. I don't have the time for this kind of project. Thanks again for your help

**Comment by DarianKimberly** [comment]
Is fine. Education is important. I quit my job because they asked me if I
valued education or my job more.

**Comment by SammyIAm** [comment]
Closing for now, good luck with school.  Feel free to re-open if you try again and have the same issue.
